src/serialadapter.c

```diff
@@ -157,10 +157,12 @@ int setport_from_serialadapter(char **portp, const SERIALADAPTER *ser, const cha
   if (1 < sa_num_matches_by_sea(ser, sp, sernum, n)) {
     pmsg_warning("-P %s is not unique; consider one of the below\n", *portp);
     for (int i = 0; i < n; i++) {
-      if (sp[i].unique && sp[i].sernum[0])
-        msg_warning("-P %s or -P %s:%s\n", sp[i].port, *portp, sp[i].sernum);
-      else
-        msg_warning("-P %s (via %s serial adapter)\n", sp[i].port, *portp);
+      if (sp[i].match) {
+        if (sp[i].unique && sp[i].sernum[0])
```

Comment:
How can `sp[i].match` and `sp[i].unique` determine whether `ser` is ambiguous wrt to the plugged-in devices? 

These flags *cannot possibly know this* as they have *never* seen `ser` at all; when they were computed in `get_libserialport_data()` that function *only* had the plugged-in devices to look at. `match` and `unique` are *not* useful for the problem of matching a *user-specified* adapter. This code cannot possibly work. It's the wrong structure.

And it does not work: I have just plugged in four devices, two ft232r with different serial number and two ch340 without serial numbers. Here the classic test `avrdude -P dunno` output, which is correct on this occasion:
```
avrdude OS error: cannot open port dunno: No such file or directory
avrdude error: unable to open programmer urclock on port dunno
Possible candidate serial ports are:
-P /dev/ttyUSB3, -P ft232r:MCU8
-P /dev/ttyUSB1 (via ch340 serial adapter)
-P /dev/ttyUSB2, -P ft232r:A600K203
-P /dev/ttyUSB0 (via ch340 serial adapter)
Note that above ports may not necessarily be connected to a target board or an AVR programmer.
Also note there may be other direct serial ports not listed above.
```

Now I test
```
$ avrdude -P ch340
avrdude warning: -P ch340 is not unique; consider one of the below
-P /dev/ttyUSB3 or -P ch340:MCU8
-P /dev/ttyUSB1 (via ch340 serial adapter)
-P /dev/ttyUSB2 or -P ch340:A600K203
-P /dev/ttyUSB0 (via ch340 serial adapter)
avrdude OS error: cannot open port ch340: No such file or directory
avrdude error: unable to open programmer urclock on port ch340
```
Half of the output is wrong on more than one level. And the same happens with
```
$ avrdude -P ft232r
avrdude warning: -P ft232r is not unique; consider one of the below
-P /dev/ttyUSB3 or -P ft232r:MCU8
-P /dev/ttyUSB1 (via ft232r serial adapter)
-P /dev/ttyUSB2 or -P ft232r:A600K203
-P /dev/ttyUSB0 (via ft232r serial adapter)
avrdude OS error: cannot open port ft232r: No such file or directory
avrdude error: unable to open programmer urclock on port ft232r
```

The reason for this is that the `match`, `unique` and `known` flags are not needed and their use yields the wrong outputs.  The current code needs some fundamental change.